lms/services/blackboard_api.py

```diff
@@ -0,0 +1,60 @@
+from lms.validation.authentication import OAuthTokenResponseSchema
+
+
+class BlackboardAPIClient:
+    def __init__(
+        self,
+        blackboard_host,
+        client_id,
+        client_secret,
+        redirect_uri,
+        http_service,
+        oauth2_token_service,
+    ):  # pylint:disable=too-many-arguments
+        self.blackboard_host = blackboard_host
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.redirect_uri = redirect_uri
+
+        self._http_service = http_service
+        self._oauth2_token_service = oauth2_token_service
+
+    def get_token(self, authorization_code):
+        """
+        Get an access token from Blackboard and save it in the DB.
+
+        :raise services.HTTPError: if something goes wrong with the access
+            token request to Blackboard
+        """
+        # Send a request to Blackboard to get an access token.
+        response = self._http_service.post(
+            f"https://{self.blackboard_host}/learn/api/public/v1/oauth2/token",
+            data={
+                "grant_type": "authorization_code",
+                "redirect_uri": self.redirect_uri,
+                "code": authorization_code,
+            },
+            auth=(self.client_id, self.client_secret),
+            schema=OAuthTokenResponseSchema,
+        )
+
+        # Save the access token to the DB.
+        self._oauth2_token_service.save(
+            response.validated_data["access_token"],
+            response.validated_data.get("refresh_token"),
+            response.validated_data.get("expires_in"),
+        )
+
+
+def factory(_context, request):
+    application_instance = request.find_service(name="application_instance").get()
+    settings = request.registry.settings
+
+    return BlackboardAPIClient(
+        blackboard_host=application_instance.lms_host(),
+        client_id=settings["blackboard_api_client_id"],
+        client_secret=settings["blackboard_api_client_secret"],
+        redirect_uri=request.route_url("blackboard_api.oauth.callback"),
+        http_service=request.find_service(name="http"),
```

Comment:
How does the HTTP service benefit from being a service rather than a plain class?

Reply:
It makes it easy for the view to get the `BlackboardAPIClient` because it can just do `request.find_service(name="blackboard_api")`. The view doesn't need to know how to construct the service, which is fairly complicated (lots of arguments, including other services).

It's nice that each service comes with its own factory method that knows how to construct it. Especially when a service depends on multiple other services as this one does. `BlackboardAPIClient`'s factory doesn't need to know how to construct `ApplicationInstanceService`, `HTTPService` or `OAuth2Service`, it just requests them.

The tests for `BlackboardAPIClient` benefit from the nice way that we have of injecting centralized mock services.

`BlackboardAPIClient` (and the other services it depends on) are all per-request singletons. So if different parts of the code need the `BlackboardAPIClient` they can each request it, you don't need to pass it around. `HTTPService` can ensure that a TCP session is re-used. `ApplicationInstanceService` can ensure that a DB query for the `ApplicationInstance` is cached. Etc.

Even though I do think there are several actual benefits to using a service here, also to be honest ervices are just what we use for these, I'm not looking for reasons to deviate. Consistency is a benefit itself.